crowdfunding/projects/views.py

```diff
@@ -1,3 +1,102 @@
 from django.shortcuts import render
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from django.http import Http404
 
-# Create your views here.
+from rest_framework import status, generics, permissions, renderers
+from .models import Project, Pledge, Comment
+from .serializers import ProjectSerializer, PledgeSerializer, ProjectDetailSerializer , PledgeDetailSerializer, CommentSerializer
+from .permissions import IsOwnerOrReadOnly, IsSupporterOrReadOnly, IsCommenterOrReadOnly
+
+from rest_framework.exceptions import NotFound
+from django.db import IntegrityError #unique = True handling
+
+from django_filters.rest_framework import DjangoFilterBackend
+
+#-- API-Root Config
+from rest_framework.decorators import api_view
+from rest_framework.reverse import reverse
+from users.models import CustomUser
+from users.serializers import CustomUserSerializer
+
+class ProjectList(generics.ListCreateAPIView):
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+    queryset = Project.objects.all()
+    serializer_class = ProjectSerializer
+    filterset_fields = ['is_open', 'owner']
+
+    def perform_create(self, serializer):
+        serializer.save(owner=self.request.user)
+        # auto adds userid as owner
+
+    def post(self, request):
+        serializer = ProjectSerializer(data=request.data)
+        # serialize data for me
+        if serializer.is_valid():
+            try:
+                serializer.save(owner=request.user)
+                return Response(serializer.data, status=status.HTTP_201_CREATED)
+            except IntegrityError:
+                #responding to unique field
+                return Response({"error":"This Project title already exists. Please enter another."}, status=status.HTTP_400_BAD_REQUEST)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+class ProjectDetail(generics.RetrieveUpdateDestroyAPIView):
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly, IsOwnerOrReadOnly]
+    queryset = Project.objects.all()
+    serializer_class = ProjectDetailSerializer
+
+    def handle_exception(self, exc):
+        if isinstance(exc, Http404):
+            return Response({'data': 'Sorry, no tree-hugging here!'},
+            status=status.HTTP_404_NOT_FOUND)
+        return super(ProjectDetail, self).handle_exception(exc)
```

Comment:
i love that you've put this in! 
it made me laugh and impressed me at the same time!
i don't think i've ever seen this used before 😍